pkg/collector/corechecks/snmp/discovery/discovery.go

```diff
@@ -0,0 +1,294 @@
+			return
+		case <-discoveryTicker.C:
+		}
+	}
+}
+
+func (d *Discovery) checkDevice(job snmpJob) error {
+	deviceIP := job.currentIP.String()
+	config := *job.subnet.config // shallow copy
+	config.IPAddress = deviceIP
+	sess, err := session.NewSession(&config)
+	if err != nil {
+		return fmt.Errorf("error configure session for ip %s: %v", deviceIP, err)
+	}
+	entityID := job.subnet.config.DiscoveryDigest(deviceIP)
+	if err := sess.Connect(); err != nil {
+		log.Debugf("subnet %s: SNMP connect to %s error: %v", d.config.Network, deviceIP, err)
+		d.deleteDevice(entityID, job.subnet)
+	} else {
+		defer sess.Close()
+
+		oids := []string{sysObjectIDOid}
+		// Since `params<GoSNMP>.ContextEngineID` is empty
+		// `params.Get` might lead to multiple SNMP GET calls when using SNMP v3
+		value, err := sess.Get(oids)
+		if err != nil {
+			log.Debugf("subnet %s: SNMP get to %s error: %v", d.config.Network, deviceIP, err)
+			d.deleteDevice(entityID, job.subnet)
+		} else if len(value.Variables) < 1 || value.Variables[0].Value == nil {
+			log.Debugf("subnet %s: SNMP get to %s no data", d.config.Network, deviceIP)
+			d.deleteDevice(entityID, job.subnet)
+		} else {
+			log.Debugf("subnet %s: SNMP get to %s success: %v", d.config.Network, deviceIP, value.Variables[0].Value)
+			d.createDevice(entityID, job.subnet, deviceIP, true)
+		}
+	}
+	return nil
+}
+
+func (d *Discovery) createDevice(entityID string, subnet *snmpSubnet, deviceIP string, writeCache bool) {
+	deviceCk, err := devicecheck.NewDeviceCheck(subnet.config, deviceIP)
+	if err != nil {
+		// should not happen since the deviceCheck is expected to be valid at this point
+		// and are only changing the device ip
+		log.Warnf("subnet %s: failed to create new device check `%s`: %s", d.config.Network, deviceIP, err)
+		return
+	}
+
+	d.discDevMu.Lock()
+	defer d.discDevMu.Unlock()
+
+	if _, present := d.discoveredDevices[entityID]; present {
+		return
+	}
+	device := Device{
+		entityID:    entityID,
+		deviceIP:    deviceIP,
+		deviceCheck: deviceCk,
+	}
+	d.discoveredDevices[entityID] = device
+	subnet.devices[entityID] = deviceIP
+	subnet.deviceFailures[entityID] = 0
+
+	if writeCache {
+		d.writeCache(subnet)
+	}
+}
+
+func (d *Discovery) deleteDevice(entityID string, subnet *snmpSubnet) {
+	d.discDevMu.Lock()
+	defer d.discDevMu.Unlock()
+	if _, present := d.discoveredDevices[entityID]; present {
+		failure, present := subnet.deviceFailures[entityID]
+		if !present {
+			subnet.deviceFailures[entityID] = 1
+			failure = 1
+		} else {
+			subnet.deviceFailures[entityID]++
+			failure++
+		}
+
+		if d.config.DiscoveryAllowedFailures != -1 && failure >= d.config.DiscoveryAllowedFailures {
+			delete(d.discoveredDevices, entityID)
+			delete(subnet.devices, entityID)
+			delete(subnet.deviceFailures, entityID)
+			d.writeCache(subnet)
+		}
+	}
+}
+
+func (d *Discovery) readCache(subnet *snmpSubnet) ([]net.IP, error) {
+	cacheValue, err := persistentcache.Read(subnet.cacheKey)
```

Comment:
cache should be protected by a mutex

Reply:
I think it's not needed since `Discovery.Start -> discoverDevices -> loadCache -> readCache`  is a call that is made when when the check initialise the first time, and there no concurrent related to this.